### ISSUE TITLE
Clean up backend BLAS/Lapack usage

### ISSUE THREAD

**Issue by RMKirkpatrick** [OpenMx/OpenMx#231]
For linear algebra, our backend uses both Eigen and BLAS/Lapack.  There exist hardware-optimized BLAS implementations (e.g., OpenBLAS, ATLAS, Intel MKL's implementation).  Further, Eigen can be compiled to use arbitrary Fortran77-compatible BLAS and Lapack implementations as its backend, and can further take advantage of additional performance boosts from other MKL features.  (As an aside, note that NPSOL and SLSQP have their own self-contained BLAS implementations, and that [NPSOL's apparently conflicts with OpenBLAS at runtime](https://openmx.ssri.psu.edu/node/4323)).

The fact that we use both BLAS/Lapack and Eigen in our backend is a barrier to making it straightforward (or perhaps even possible) to compile OpenMx with a machine-optimized BLAS/Lapack implementation.  That's because the code is currently written to use R's BLAS & Lapack API, so trying to compile against a different BLAS/Lapack implementation introduces conflicting definitions of the subroutines.  It's also not as simple as just dropping R's BLAS/Lapack API, because then the `F77_CALL()`s to the subroutines aren't defined.

Some of our BLAS/Lapack calls are done via inlined wrapper functions, such as those defined in src/omxMatrix.h :
```
static OMXINLINE void omxDAXPY(double alpha, omxMatrix* lhs, omxMatrix* rhs) {              // RHS += alpha*lhs  
    // N.B.  Not fully tested.                                                              // Assumes common majority or vectordom.
    if(lhs->colMajor != rhs->colMajor) { omxToggleRowColumnMajor(rhs);}
    int len = lhs->rows * lhs->cols;
    int onei = 1;
    F77_CALL(daxpy)(&len, &alpha, lhs->data, &onei, rhs->data, &onei);

}
```
Other times, we directly use the `F77_CALL()`.

I propose the following:
1.  Replace all direct Fortran BLAS/Lapack calls with wrappers; this may require writing a few new wrappers.
2.  Condition the definition of the wrappers on a compile-time option.  The default will be to use R's API.
3.  Also condition which BLAS & Lapack headers to #include on that compile-time option.

For example, `omxDAXPY()` could be redefined as follows:
```
static OMXINLINE void omxDAXPY(double alpha, omxMatrix* lhs, omxMatrix* rhs) {              // RHS += alpha*lhs  
    // N.B.  Not fully tested.                                                              // Assumes common majority or vectordom.
    if(lhs->colMajor != rhs->colMajor) { omxToggleRowColumnMajor(rhs);}
    int len = lhs->rows * lhs->cols;
    int onei = 1;
#ifdef MX_USE_MKL
    daxpy(*len, *alpha, lhs->data, *onei, rhs->data, *onei);
#endif
#ifndef MX_USE_MKL
    F77_CALL(daxpy)(&len, &alpha, lhs->data, &onei, rhs->data, &onei);
#endif

}
```
Thoughts?

**Comment by jpritikin** [comment]
I really think we should try to avoid adding conditional code. If Eigen can wrap BLAS then can we replace all BLAS calls with Eigen?

**Comment by RMKirkpatrick** [comment]
> If Eigen can wrap BLAS then can we replace all BLAS calls with Eigen?

I suppose that's another possibility.

> I really think we should try to avoid adding conditional code.

I understand, but I don't see how we can avoid that if we want to give users a choice, at minimum, of which BLAS/Lapack headers to #include at compiletime.

**Comment by jpritikin** [comment]
> I understand, but I don't see how we can avoid that if we want to give users a choice, at minimum, of which BLAS/Lapack headers to #include at compile time.

Switching between 2 include files seems like a very modest use of conditional code. I'm asking to minimize, acknowledging that elimination may not be desirable.

**Comment by jpritikin** [comment]
>   If Eigen can wrap BLAS then can we replace all BLAS calls with Eigen?

CSOLNP already uses Eigen. What about `slsqp.c`? Would we need to convert this code somehow?

**Comment by RMKirkpatrick** [comment]
> What about `slsqp.c`? Would we need to convert this code somehow?

I don't believe so.  The BLAS code in that file seems intended to be local to SLSQP.  When I (think I) compiled OpenMx against the MKL BLAS and Lapack earlier this week, I didn't have to change anything in slsqp.c.

**Comment by jpritikin** [comment]
If I grep for F77_CALL then I find,

```
bvnd)
daxpy)
ddot)
dgeev)
dgels)
dgemm)
dgemv)
dgesdd)
dgesv)
dgetrf)
dgetri)
dpotrf)
dpotri)
dsyevr)
dsymm)
dsymv)
dsytrf)
dsytri)
omxunsafedgemm)
omxunsafedgemv)
sadmvn)
npoptn)
npsol)
```
That's not very many functions to replace. `npoptn` and `npsol` are part of NPSOL. What about `sadmvn`? Do we need to change anything there?

**Comment by RMKirkpatrick** [comment]
I don't think anything about sadmvn needs to change.

**Comment by RMKirkpatrick** [comment]
As of 8/15/19, @jpritikin replaced all Fortran BLAS/LAPACK calls in the OpenMx backend with equivalent code that uses Eigen.  The OpenMx backend used to have an omxBLAS.f, which contained "unsafe" implementations of the BLAS/LAPACK subroutines we use.  Those subroutines are "unsafe" because, contrary to IEEE standard, they don't necessarily propagate `NaN`s and other non-finite values through multiplication.  Their lack of "safety", however, made them faster in some circumstances.

We scrapped src/omxBLAS.f because CRAN required that we stop passing strings to Fortran subroutines.  But that change broke something subtle with omxBLAS.f, which Joshua couldn't figure out.  Since our backend now no longer uses any calls to Fortran BLAS/LAPACK subroutines,  I suggest we re-implement the "unsafety" in Joshua's new Eigen code.  By default, it will be switched off, but may be switched on by the user through a runtime option.

**Comment by jpritikin** [comment]
>  Since our backend now no longer uses any calls to Fortran BLAS/LAPACK subroutines, I suggest we re-implement the "unsafety" in Joshua's new Eigen code. By default, it will be switched off, but may be switched on by the user through a runtime option.

Wouldn't this optimization be better implemented in Eigen itself? We'd have to contact Eigen developers and see whether they were interested in such a feature.

**Comment by RMKirkpatrick** [comment]
> Wouldn't this optimization be better implemented in Eigen itself?  We'd have to contact Eigen developers and see whether they were interested in such a feature.

I kind of doubt it.  For one thing, they may not be receptive.  Plus, I looked back at the omxBLAS.f code, which only contained 2 subroutine implementations: DGEMM and DGEMV.  The "unsafe" conditional shortcuts the code takes make the most sense in the context of this kind of subroutine, e.g. if scalar `beta` is zero, then the product of `beta` and `C` is a matrix of zeroes.

Incidentally, multiplying a matrix by a zero-valued scalar is a foreseeably common use case with mxGREML, arising when a variance component has been fixed to zero to fit a "reduced" model, to be compared with a "full" model.
